packages/native/src/main.udl

```diff
@@ -140,47 +90,58 @@ interface FFIExtendableUriResolver {
 };
 
 interface FFIRecursiveUriResolver {
-  constructor(FFIUriResolverLike uri_resolver_like);
+  constructor(FFIUriResolver uri_resolver);
 };
 
-interface FFIClient {
-  [Throws=Error]
-  sequence<u8> invoke_raw(Uri uri, [ByRef] string method, sequence<u8>? args, string? env);
+dictionary FFIUriResolutionStep {
+  FFIUri source_uri;
+  FFIUriPackageOrWrapper result;
+  string? description;
+  sequence<FFIUriResolutionStep>? sub_history;
+};
 
-  [Throws=Error]
-  sequence<Uri> get_implementations(Uri uri);
+interface FFIUriResolutionContext {
+  constructor();
+  void set_resolution_path(sequence<string> resolution_path);
+  void set_history(sequence<FFIUriResolutionStep> history);
+  void set_resolving_uri_map(record<DOMString, boolean> resolving_uri_map);
+  void set_start_resolving(FFIUri uri);
+  void set_stop_resolving(FFIUri uri);
+  void track_step(FFIUriResolutionStep step);
+  sequence<FFIUriResolutionStep> get_history();
+  sequence<FFIUri> get_resolution_path();
+  FFIUriResolutionContext create_sub_history_context();
+  FFIUriResolutionContext create_sub_context();
+};
+
+interface FFIClient {
+  [Throws=FFIError]
+  sequence<u8> invoke_raw(FFIUri uri, [ByRef] string method, sequence<u8>? args, sequence<u8>? env, FFIUriResolutionContext? resolution_context);
 
-  record<DOMString, sequence<Uri>>? get_interfaces();
+  [Throws=FFIError]
+  sequence<FFIUri> get_implementations(FFIUri uri);
 
-  string? get_env_by_uri(Uri uri);
+  record<DOMString, sequence<FFIUri>>? get_interfaces();
 
-  [Throws=Error]
-  sequence<u8> invoke_wrapper_raw(FFIWrapper wrapper, Uri uri, [ByRef] string method, sequence<u8>? args, string? env);
+  sequence<u8>? get_env_by_uri(FFIUri uri);
 
-  [Throws=Error]
-  FFIWrapper load_wrapper(Uri uri);
+  [Throws=FFIError]
+  sequence<u8> invoke_wrapper_raw(FFIWrapper wrapper, FFIUri uri, [ByRef] string method, sequence<u8>? args, sequence<u8>? env, FFIUriResolutionContext? resolution_context);
```

Comment:
Do we need this right now if we don't have load_wrapper? I guess someone could invoke a externally loaded wrapper, but the use case is very minor

Reply:
True. Leant towards exposing `load_wrapper` instead of removing this